### PR TITLE
Fix for Issue #77

### PR DESCRIPTION
**Pull request by MaryBennett**
2 commits:
    Rename bitmanip constraints
    
    Files Changed:
    
     * constraints.md: Change contraint names.
     * corev.md: Likewise.

  Fix for Issue #77
    
    Since the constraint `m` should assume non-xcvmem specific operands,
    we redefined TARGET_MEM_CONSTRAINT and defined 'm' to exclude XCVmem
    specific operands.
    Constraint `CVmp` defines addresses for post modify operands.
    Constraint `CVmr` defines addresses for reg + reg operands.
    
    Files Changed:
    
       * gcc/config/riscv/predicates.md (mem_post_inc): Prevent DI and
         DF mode mem.
         (mem_plus_reg): Likewise.
       * gcc/config/riscv/constraints.md: Add new constraints `m`,
         `CVmp` and `CVmr`.
       * gcc/config/riscv/riscv.md: Remove constraint `am`.
       * gcc/config/riscv/riscv.h: Redefine TARGET_MEM_CONSTRAINT to
         `w`.
       * gcc/testsuite/gcc.target/riscv/cv-mem-compile-1.c: New test.
       * gcc/testsuite/gcc.target/riscv/cv-mem-compile-2.c: Likewise.
       * gcc/testsuite/gcc.target/riscv/cv-mem-compile-3.c: Likewise.

### CODE REVIEW

**Comment by MaryBennett** [comment]
```
| Category             | Previous       | Current       |       Delta |
| :------------------- | -------------: | ------------: | ----------: |
| Expected passes      |          19143 |         19178 |         +35 |
| Unexpected failures  |            173 |           173 |           - |
| Unexpected successes |              - |             - |           - |
| Expected failures    |              6 |             6 |           - |
| Unresolved testcases |              6 |             6 |           - |
| Unsupported tests    |           1277 |          1280 |          +3 |
```

Additional unsupported due to O0 tests marked unsupported.

File: gcc/config/riscv/constraints.md [hunk 1]
```diff
@@ -129,9 +129,12 @@
        (match_test "GET_CODE(XEXP(op,0)) == REG")))
 
 (define_memory_constraint "am"
-  "An address that is held in a general-purpose register."
+  "An address that is not for PLUS reg reg."
```
Comment (craigblackmore) [inline gcc/config/riscv/constraints.md]: An address that is not base reg + index reg

File: gcc/config/riscv/constraints.md [hunk 2]
```diff
@@ -129,9 +129,12 @@
...
   (and (match_code "mem")
-       (match_test "!(GET_CODE(XEXP(op,0)) == PLUS && GET_CODE(XEXP(XEXP(op,0),0)) == REG && GET_CODE(XEXP(XEXP(op,0),1)) == REG)")))
+       (match_test "!(GET_CODE (XEXP (op, 0)) == PLUS
+			&& GET_CODE (XEXP (XEXP (op, 0), 0)) == REG
+			&& GET_CODE (XEXP (XEXP (op, 0), 1)) == REG)
+		    && GET_CODE (XEXP (op, 0)) != POST_MODIFY")))
```
Comment (craigblackmore) [inline gcc/config/riscv/constraints.md]: The description of this constraint doesn't mention anything about post modify.

File: gcc/config/riscv/predicates.md [hunk 3]
```diff
@@ -236,11 +236,15 @@
 
 (define_predicate "mem_post_inc"
   (and (match_code "mem")
-       (match_test "TARGET_XCVMEM && GET_CODE (XEXP (op, 0)) == POST_MODIFY")))
+       (match_test "TARGET_XCVMEM && GET_CODE (XEXP (op, 0)) == POST_MODIFY
+		    && GET_MODE (op) != DImode
+                    && GET_MODE (op) != DFmode")))
```
Comment (craigblackmore) [inline gcc/config/riscv/predicates.md]: Is it DI / DF specifically that you want to exclude or modes larger than a certain size? What about TI / TF for example?

File: gcc/config/riscv/riscv.md [hunk 4]
```diff
@@ -1718,7 +1718,7 @@
 (define_insn "zero_extendqi<SUPERQI:mode>2"
   [(set (match_operand:SUPERQI 0 "register_operand"    "=r,r")
 	(zero_extend:SUPERQI
-	    (match_operand:QI 1 "nonimmediate_nonpostinc" " r,m")))]
+	    (match_operand:QI 1 "nonimmediate_nonpostinc" " r,am")))]
```
Comment (craigblackmore) [inline gcc/config/riscv/riscv.md]: Selectively changing some of the `m`'s to `am` seems haphazard. I think base reg + index reg and post modify should be the exceptions and `m` the norm. Have a look at `TARGET_MEM_CONSTRAINT`.

File: gcc/config/riscv/riscv.md [hunk 5]
```diff
@@ -1813,8 +1813,8 @@
 })
 
 (define_insn "*movhf_hardfloat"
-  [(set (match_operand:HF 0 "nonimmediate_nonpostinc" "=f,   f,f,f,m,m,*f,*r,  *r,*r,*m")
-	(match_operand:HF 1 "move_operand"         " f,zfli,G,m,f,G,*r,*f,*G*r,*m,*r"))]
+  [(set (match_operand:HF 0 "nonimmediate_nonpostinc" "=f,   f,f, f,am,m,*f,*r,  *r,*r,*m")
+	(match_operand:HF 1 "move_operand"            " f,zfli,G,am, f,G,*r,*f,*G*r,*m,*r"))]
```
Comment (craigblackmore) [inline gcc/config/riscv/riscv.md]: Unrelated whitespace change here and a few times below.

File: gcc/testsuite/gcc.target/riscv/cv-mem-compile-1.c [hunk 6]
```diff
@@ -0,0 +1,14 @@
+/* { dg-do compile } */
```
Comment (craigblackmore) [inline gcc/testsuite/gcc.target/riscv/cv-mem-compile-1.c:1]: For each test, add a comment stating what is being tested.

**Review by craigblackmore — CHANGES_REQUESTED** [review]

File: gcc/config/riscv/predicates.md (hunk 3 above)
Comment (MaryBennett) [reply gcc/config/riscv/predicates.md]: Good idea

File: gcc/config/riscv/riscv.md (hunk 5 above)
Comment (MaryBennett) [reply gcc/config/riscv/riscv.md]: This was so the formatting matched upstream. Should I revert it?

File: gcc/config/riscv/constraints.md [hunk 7]
```diff
@@ -35,6 +35,14 @@
 
 ;; General constraints
 
+(define_memory_constraint "m"
+  "An address that is not base reg + index reg or post modify."
+  (and (match_code "mem")
+       (match_test "!(GET_CODE (XEXP (op, 0)) == PLUS
+                        && GET_CODE (XEXP (XEXP (op, 0), 0)) == REG
+                        && GET_CODE (XEXP (XEXP (op, 0), 1)) == REG)
+                    && GET_CODE (XEXP (op, 0)) != POST_MODIFY")))
```
Comment (craigblackmore) [inline gcc/config/riscv/constraints.md]: The default `m` constraint does more than just check if code is `mem` https://github.com/openhwgroup/corev-gcc/blob/4a6bad0f495f038f948a00a40464e3d28761b0e4/gcc/common.md#L24-L27

You can also reuse the new CVmp constraint.

Check if this works and is sufficient:
```
(and (match_operand 0 "TARGET_MEM_CONSTRAINT")
      not (match_operand 0 "CVmp"))
```
Hopefully `"TARGET_MEM_CONSTRAINT"` works, I've based it on how it's used in `common.md`.

File: gcc/config/riscv/constraints.md [hunk 8]
```diff
@@ -168,6 +171,23 @@
        (and (match_test "IN_RANGE (ival, 0, 1073741823)")
             (match_test "exact_log2 (ival + 1) != -1"))))
 
+(define_constraint "CVmp"
+  "An address for post-modify or reg+reg stores and loads"
+  (and (match_code "mem")
```
Comment (craigblackmore) [inline gcc/config/riscv/constraints.md]: Again, instead of `match_code "mem"`, reuse the default mem constraint `(match_operand 0 "TARGET_MEM_CONSTRAINT")`.

File: gcc/config/riscv/constraints.md [hunk 9]
```diff
@@ -168,6 +171,23 @@
...
+       (match_test "GET_CODE (XEXP (op, 0)) == POST_MODIFY
+		|| (GET_CODE (XEXP (op, 0)) == PLUS
+		&& GET_CODE (XEXP (XEXP (op, 0), 0)) == REG
+                && GET_CODE (XEXP (XEXP (op, 0), 1)) == REG)"))
```
Comment (craigblackmore) [inline gcc/config/riscv/constraints.md]: You could also reuse the `CVmr` constraint here.

File: gcc/config/riscv/constraints.md [hunk 10]
```diff
@@ -168,6 +171,23 @@
...
+)
+
+(define_constraint "CVmr"
+  "An address for reg+reg stores and loads"
+  (and (match_code "mem")
```
Comment (craigblackmore) [inline gcc/config/riscv/constraints.md:178]: Again, instead of `match_code "mem"`, reuse the default mem constraint `(match_operand 0 "TARGET_MEM_CONSTRAINT")`.

File: gcc/config/riscv/corev.md [hunk 11]
```diff
@@ -3096,23 +3096,23 @@
 ;; Normal Register-Register Load/Store
 (define_insn "cv_load<mode>"
    [(set (match_operand:ANYI 0 "register_operand" "=r")
-         (match_operand:ANYI 1 "mem_plus_reg" "m"))]
+         (match_operand:ANYI 1 "mem_plus_reg" "CVmp"))]
```
Comment (craigblackmore) [inline gcc/config/riscv/corev.md]: Should this be using the `CVmr` constraint? It's under the reg-reg comment.

File: gcc/config/riscv/predicates.md [hunk 12]
```diff
@@ -236,11 +236,19 @@
...
+                    && GET_MODE (op) != TFmode")))
 
 (define_predicate "mem_plus_reg"
   (and (match_code "mem")
-       (match_test "GET_CODE (XEXP (op, 0)) == PLUS
+       (match_test "TARGET_XCVMEM && GET_CODE (XEXP (op, 0)) == PLUS
+		    && GET_MODE (op) != DImode
+		    && GET_MODE (op) != DFmode
+		    && GET_MODE (op) != TImode
+		    && GET_MODE (op) != TFmode
```
Comment (craigblackmore) [inline gcc/config/riscv/predicates.md]: `GET_MODE_SIZE (op) > GET_MODE_SIZE (SImode)`? Otherwise check for the exact modes which _are_ supported.

File: gcc/config/riscv/constraints.md (hunk 8 above)
Comment (MaryBennett) [reply gcc/config/riscv/constraints.md]: reference to unknown predicate 'TARGET_MEM_CONSTRAINT'
I can't do this.

Comment (craigblackmore) [reply gcc/config/riscv/constraints.md]: I should have said `(match_operand 0 "memory_operand" "TARGET_MEM_CONSTRAINT")`, but then the predicate gets checked which is unnecessary and wasn't being done before. I'd just copy the conditions from the default in `common.md` then.

File: gcc/config/riscv/constraints.md (hunk 7 above)
Comment (craigblackmore) [reply gcc/config/riscv/constraints.md]: Actually, could you reuse the cv mem predicates e.g.
```
(not (match_operand 0 "mem_post_inc"))
```
and
```
(not (match_operand 0 "mem_plus_reg"))
```
